src/recon_buildblock/LogcoshPrior.cxx

```diff
@@ -0,0 +1,666 @@
+        sprintf(filename, "%s%d.v", gradient_filename_prefix.c_str(), count);
+        OutputFileFormat<DiscretisedDensity<3,elemT> >::default_sptr()->
+                write_to_file(filename, prior_gradient);
+        delete[] filename;
+      }
+    }
+
+    template <typename elemT>
+    void
+    LogcoshPrior<elemT>::
+    compute_Hessian(DiscretisedDensity<3,elemT>& prior_Hessian_for_single_densel,
+                    const BasicCoordinate<3,int>& coords,
+                    const DiscretisedDensity<3,elemT> &current_image_estimate)
+    {
+      assert(  prior_Hessian_for_single_densel.has_same_characteristics(current_image_estimate));
+      prior_Hessian_for_single_densel.fill(0);
+      if (this->penalisation_factor==0)
+      {
+        return;
+      }
+
+
+      const DiscretisedDensityOnCartesianGrid<3,elemT>& current_image_cast =
+              dynamic_cast< const DiscretisedDensityOnCartesianGrid<3,elemT> &>(current_image_estimate);
+
+      DiscretisedDensityOnCartesianGrid<3,elemT>& prior_Hessian_for_single_densel_cast =
+              dynamic_cast<DiscretisedDensityOnCartesianGrid<3,elemT> &>(prior_Hessian_for_single_densel);
+
+      if (weights.get_length() ==0)
+      {
+        compute_weights(weights, current_image_cast.get_grid_spacing(), this->only_2D);
+      }
+
+
+      const bool do_kappa = !is_null_ptr(kappa_ptr);
+
+      if (do_kappa && kappa_ptr->has_same_characteristics(current_image_estimate))
+        error("LogcoshPrior: kappa image has not the same index range as the reconstructed image\n");
+
+      const int z = coords[1];
+      const int y = coords[2];
+      const int x = coords[3];
+      const int min_dz = max(weights.get_min_index(), prior_Hessian_for_single_densel.get_min_index()-z);
+      const int max_dz = min(weights.get_max_index(), prior_Hessian_for_single_densel.get_max_index()-z);
+
+      const int min_dy = max(weights[0].get_min_index(), prior_Hessian_for_single_densel[z].get_min_index()-y);
+      const int max_dy = min(weights[0].get_max_index(), prior_Hessian_for_single_densel[z].get_max_index()-y);
+
+      const int min_dx = max(weights[0][0].get_min_index(), prior_Hessian_for_single_densel[z][y].get_min_index()-x);
+      const int max_dx = min(weights[0][0].get_max_index(), prior_Hessian_for_single_densel[z][y].get_max_index()-x);
+
+      elemT diagonal = 0;
+      for (int dz=min_dz;dz<=max_dz;++dz)
+        for (int dy=min_dy;dy<=max_dy;++dy)
+          for (int dx=min_dx;dx<=max_dx;++dx)
+          {
+            // dz==0,dy==0,dx==0 will have weight 0, so we can just include it in the loop
+            elemT current =
+                    weights[dz][dy][dx];
+
+            if (do_kappa)
+              current *=
+                      (*kappa_ptr)[z][y][x] * (*kappa_ptr)[z+dz][y+dy][x+dx];
+
+            diagonal += current;
+            prior_Hessian_for_single_densel_cast[z+dz][y+dy][x+dx] = -current*this->penalisation_factor;
```

Comment:
this seems a copy of the QP code. If we don't need it, you could throw an error (and say so in the doxygen)

Reply:
Logcosh does have a second derivitive
`(x sech^2(|x|/scalar)) / (|x|)`, where x is `temp = x_i - x_j`. I think we add it. This is actually a nice and well behaved function.